libbeat/processors/ratelimit/docs/rate_limit.asciidoc

```diff
@@ -38,3 +41,4 @@ The following settings are supported:
 
 `limit`:: The rate limit. Supported time units for the rate are `s` (per second), `m` (per minute), and `h` (per hour).
 `fields`:: (Optional) List of fields. The rate limit will be applied to each distinct value derived by combining the values of these fields.
+`metric_field`:: (Optional) When events are rate limited, the first event that is allowed to pass through after rate limiting will contain this field. The value of the field will be the number of events that were rate limited since the last time an event was allowed to pass through.
```

Comment:
I wonder if we should decide what field to use for this. This way users wouldn't need to think what field to use, with the risk of choosing a field that is used by something else. We could also provide a mapping for the field. And we could also keep this always enabled.

(Naming is complicated, I don't have a suggestion for this field :grimacing: )

As a example of something somehow similar, the `multiline` feature of `filebeat` adds a `multiline` flag in a known field (`log.flags`) when multiple lines are merged in a single event, and this cannot be disabled.

Reply:
I recall having a conversation previously where we thought it might be a good idea to give the user control of the field: https://github.com/elastic/beats/issues/21020#issuecomment-742468061. But your points here about the advantages of having a fixed field are also valid. 

Maybe we should take a step back and ask what is the use case for including a field in the event that describes a past state, especially considering that events may be batched and retried by some outputs. I think the idea is to know that rate limiting is happening or not and, if it is, to what extent. So maybe a monitoring counter is the more reliable way of expressing this?

Reply:
> I think the idea is to know that rate limiting is happening or not and, if it is, to what extent.

Yes, I think the same.

> So maybe a monitoring counter is the more reliable way of expressing this?

A monitoring counter sounds good, but I wonder if this is enough when multiple `rate_limit` processors are configured in the same beat, it can be difficult to see which one is being triggered. Though maybe this is not such a problem in real deployments.
Having a counter for each `rate_limit` could be an alternative, but it can complicate things, and I am not sure how each counter could be correlated to each config.
Having the info in the event is not so nice or reliable, but I guess it is easier to identify what kind of events are being throttled.

Maybe by now we can go on with a global monitoring counter (one for all `rate_limit` processors), and wait for feedback to see if this is enough. Having debug logging about the throttled events could also help with tricky cases and complicated configs.

Reply:
> A monitoring counter sounds good, but I wonder if this is enough when multiple rate_limit processors are configured in the same beat, it can be difficult to see which one is being triggered. Though maybe this is not such a problem in real deployments.

I looked at how the `dns` processor solves this issue. It does so by giving each _instance_ of the processor it's own ID and then using it for logging and metrics monitoring purposes:

https://github.com/elastic/beats/blob/11c5367b33fb0a92d93c24fb6d382c1795aed5a5/libbeat/processors/dns/dns.go#L60-L65

So maybe we could play the same game here?

Alternatively we could create a new `processors.MonitoredProcessor` struct that implements `processors.Processor` but also implements a `MonitoringRegistry()` method that provides a monitoring registry instance for any processor that wishes to be monitored. This way we will have a consistent namespace for each processor instance within the libbeat monitoring registry for per-processor-instance metrics.

WDYT?

> Having debug logging about the throttled events could also help with tricky cases and complicated configs.
Agreed. We already have this today: https://github.com/elastic/beats/blob/11c5367b33fb0a92d93c24fb6d382c1795aed5a5/libbeat/processors/ratelimit/rate_limit.go#L89

Reply:
Ok, if we have already other processors doing something like this I am ok with following the same approach :+1:

Reply:
Cool, I'm going to change this PR to remove the metric field on the event and instead setup a monitoring counter for the cumulative total number of events rate limited.